rfc/0052-sdk3-full-text-search.md

```diff
@@ -566,7 +566,279 @@ A query that matches nothing.
+It can contain options applicable just to that vector query such as "boost" and "k".
+* A way to perform multiple `VectorQuery`s, which we can call a `VectorSearch`.
+Maps to the top-level `knn` array.
+* Any options applicable to all forms of query (both the `SearchQuery` and the `VectorSearch`), which are in `SearchOptions` (skip, limit, explain, highlight, fields, etc.)  We can continue to use the existing `SearchOptions` object for this.
+* Any options applicable only to the `VectorSearch`, which we can call `VectorSearchOptions`.
+
+So the user will be performing a `SearchRequest` containing a `SearchQuery` and/or a `VectorSearch` and/or a `FutureSearchType`, optionally with some `SearchOptions` applying to all of those.  
+A `VectorSearch` comprises 1+ `VectorQuery`s and optionally some `VectorSearchOptions`.
+
+Neither `VectorSearch` nor `VectorQuery` will extend the `SearchQuery` interface, which we will reserve for traditional FTS queries.
+
+Hence, none of this is compatible with the existing search interface `cluster.searchQuery(SearchQuery, SearchOptions)`.
+
+This takes us to a new API, `cluster.search(String indexName, SearchRequest, SearchOptions)`, which has the flexibility to be able to perform traditional FTS and/or vector queries, and/or any future top-level queries.
+
+The current `cluster.searchQuery()` API will not be deprecated at this time, given the volatility of this new feature.
+But the intent is that the new `cluster.search()` will ultimately be how new users perform any sort of FTS service query, including just a standard FTS `SearchQuery` without vector search.
+
+### API Examples
+All examples will use the reference Java SDK implementation.
+
+1. A traditional FTS query without vector search.  With the existing API:
+
+```java
+SearchQuery query = SearchQuery.matchAll();
+
+scope.searchQuery("search_index_name", query);
+cluster.searchQuery("search_index_name", query);
+```
+
+and with the new API:
+
+```java
+SearchRequest request = SearchRequest.searchQuery(SearchQuery.matchAll());
+
+cluster.search("search_index_name", request);
+scope.search("search_index_name", request);
+```
+
+2. A vector search with a single vector query and no traditional FTS query:
+
+```java
+// Using an imaginary API provided by OpenAI to generate the vector.
+float[] vector = OpenAI.createVectorFor("some query");
+
+SearchRequest request = SearchRequest
+        .vectorSearch(VectorSearch.create(VectorQuery.create("vector_field", vector)));
+
+scope.search("search_index_name", request);
+cluster.search("search_index_name", request);
+```
+
+3. A combined search with both vector search and a traditional FTS query:
+
+```java
+float[] vector = OpenAI.createVectorFor("some query");
+
+SearchRequest request = SearchRequest
+        .searchQuery(SearchQuery.matchAll())
+        .vectorSearch(VectorSearch.create(VectorQuery.create("vector_field", vector)));
+
```

Comment:
Are we attached to `searchQuery` for historical reasons (as in what we have today so less ).It would be ideal to use` ftsSearch` and `ftsQuery` to align with `vectorSearch`/`vectorQuery` respectively- no?
A new developer reading the API without any of the context would find it hard to understand that searchQuery is actually FTS.

Reply:
Possibly - Ray and I have certainly wondered before why it's named "search" and not "fullTextSearch" or similar - but it potentially opens a rather large can of worms.  `.ftsSearch(SearchQuery.matchAll())` or `.fullTextSearch(SearchQuery.matchAll())` both look weirdly inconsistent to me - would probably want something instead like `.fullTextSearch(FullTextSearch[Query].matchAll())`, which looks cleaner.  But we have a lot of classes implementing that `SearchQuery` interface, so that's potentially an expensive addition - some SDKs may end up having to C&P a bunch.

Reply:
+1 to what @rajagp said.n It would be great to have consistency in function names ftsQuery, geoDistance, vectorQuery, etc.

Reply:
For the existing SearchQuery including geo I've used the existing names so the large corpus of existing SDK code based on it can be refused. Plus it is somewhat orthogonal to the vector search work, and I didn't want to unduly scope creep given the timescales required.

It's worth noting that this proposal was originally presented for PM review last year, and we are due to begin implementing tomorrow. I can still make small bug fixes but it is too late for feedback that will require such large ranging changes - unless everything is pushed back by a sprint. Let me know if you want to go that path; I'm happy to discuss renaming if given time to do it, though will need details on the desired names from PMs.